src/Features/Core/Portable/Diagnostics/EngineV2/CompilerDiagnosticExecutor.cs

```diff
@@ -0,0 +1,292 @@
+
+                    var semantic = await analyzerDriver.GetAnalyzerSemanticDiagnosticsAsync(model, noSpanFilter, oneAnalyzers, cancellationToken).ConfigureAwait(false);
+                    result.AddSemanticDiagnostics(tree, CompilationWithAnalyzers.GetEffectiveDiagnostics(semantic, compilation));
+                }
+
+                var rest = await analyzerDriver.GetAnalyzerCompilationDiagnosticsAsync(oneAnalyzers, cancellationToken).ConfigureAwait(false);
+                result.AddCompilationDiagnostics(CompilationWithAnalyzers.GetEffectiveDiagnostics(rest, compilation));
+
+                builder.Add(analyzer, result.ToResult());
+            }
+
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// We have this builder to avoid creating collections unnecessarily.
+        /// Expectation is that, most of time, most of analyzers doesn't have any diagnostics. so no need to actually create any objects.
+        /// </summary>
+        internal struct Builder
+        {
+            private readonly Project _project;
+            private readonly VersionStamp _version;
+
+            private HashSet<DocumentId> _lazySet;
+
+            private Dictionary<DocumentId, List<DiagnosticData>> _lazySyntaxLocals;
+            private Dictionary<DocumentId, List<DiagnosticData>> _lazySemanticLocals;
+            private Dictionary<DocumentId, List<DiagnosticData>> _lazyNonLocals;
+
+            private List<DiagnosticData> _lazyOthers;
+
+            public Builder(Project project, VersionStamp version)
+            {
+                _project = project;
+                _version = version;
+
+                _lazySet = null;
+                _lazySyntaxLocals = null;
+                _lazySemanticLocals = null;
+                _lazyNonLocals = null;
+                _lazyOthers = null;
+            }
+
+            public AnalysisResult ToResult()
+            {
+                var documentIds = _lazySet == null ? ImmutableHashSet<DocumentId>.Empty : _lazySet.ToImmutableHashSet();
+                var syntaxLocals = Convert(_lazySyntaxLocals);
+                var semanticLocals = Convert(_lazySemanticLocals);
+                var nonLocals = Convert(_lazyNonLocals);
+                var others = _lazyOthers == null ? ImmutableArray<DiagnosticData>.Empty : _lazyOthers.ToImmutableArray();
+
+                return new AnalysisResult(_project.Id, _version, syntaxLocals, semanticLocals, nonLocals, others, documentIds);
+            }
+
+            private ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> Convert(Dictionary<DocumentId, List<DiagnosticData>> map)
+            {
+                return map == null ? ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty : map.ToImmutableDictionary(kv => kv.Key, kv => kv.Value.ToImmutableArray());
+            }
+
+            public void AddExternalSyntaxDiagnostics(DocumentId documentId, IEnumerable<Diagnostic> diagnostics)
```

Comment:
Can you add a comment on what you mean by **External** diagnostics here and what is the source of these diagnostics? Looking at the callsite it seems like those coming from projects which don't support compilation.